core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -160,55 +172,74 @@ public void clusterChanged(final ClusterChangedEvent event) {
-                        logger.debug("{}[{}] removing shard (disabled block persistence)", index, shardId);
-                        try {
-                            indexService.removeShard(shardId, "removing shard (disabled block persistence)");
-                        } catch (Throwable e) {
-                            logger.warn("{} failed to remove shard (disabled block persistence)", e, index);
-                        }
-                    }
-                    removeIndex(index, "cleaning index (disabled block persistence)");
+                for (AllocatedIndex<? extends Shard> indexService : indicesService) {
+                    indicesService.removeIndex(indexService.index(), "cleaning index (disabled block persistence)"); // also cleans shards
                 }
                 return;
             }
 
-            cleanFailedShards(event);
-
-            // cleaning up indices that are completely deleted so we won't need to worry about them
-            // when checking for shards
-            applyDeletedIndices(event);
-            applyDeletedShards(event);
-            // call after deleted shards so indices with no shards will be cleaned
-            applyCleanedIndices(event);
-            // make sure that newly created shards use the latest meta data
-            applyIndexMetaData(event);
-            applyNewIndices(event);
-            // apply mappings also updates new indices. TODO: make new indices good to begin with
-            applyMappings(event);
-            applyNewOrUpdatedShards(event);
+            updateFailedShardsCache(event);
+
+            deleteIndices(event); // also deletes shards of deleted indices
+
+            removeIndices(event); // also removes shards of removed indices
+
+            removeShards(event);
+
+            updateIndices(event); // can also fail shards, but these are then guaranteed to be in failedShardsCache
+
+            createIndices(event);
+
+            createOrUpdateShards(event);
         }
     }
 
-    private void cleanFailedShards(final ClusterChangedEvent event) {
-        RoutingNode routingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
-        if (routingNode == null) {
-            failedShards.clear();
+    /**
+     * Removes shard entries from the failed shards cache that are no longer allocated to this node by the master.
+     * Sends shard failures for shards that are marked as actively allocated to this node but don't actually exist on the node.
+     * Resends shard failures for shards that are still marked as allocated to this node but previously failed.
+     *
+     * @param event cluster change event
+     */
+    private void updateFailedShardsCache(final ClusterChangedEvent event) {
+        RoutingNode localRoutingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
+        if (localRoutingNode == null) {
```

Comment:
I presume changing routing nodes to always have an entry for all nodes will be a follow up?